// lerna-changelog/lib/progress-bar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var chalk_1 = require("chalk");
var ProgressBar = require("progress");
var padEnd = require("string.prototype.padend");
var ProgressBarController = (function () {
    function ProgressBarController() {
        this.bar = null;
    }
    ProgressBarController.prototype.init = function (title, total) {
        if (this.bar) {
            this.terminate();
        }
        if (!process.stdout.isTTY) {
            return;
        }
        this.bar = new ProgressBar(":bar " + title + " (:percent)", {
            total: total,
            complete: chalk_1.default.hex("#0366d6")("█"),
            incomplete: chalk_1.default.enabled ? chalk_1.default.gray("█") : "░",
            clear: true,
            width: 20,
        });
    };
    ProgressBarController.prototype.tick = function () {
        if (this.bar) {
            this.bar.tick(1);
        }
    };
    ProgressBarController.prototype.clear = function () {
        if (this.bar) {
            this.bar.terminate();
        }
    };
    ProgressBarController.prototype.terminate = function () {
        this.clear();
        this.bar = null;
    };
    return ProgressBarController;
}());
exports.default = new ProgressBarController();
